File: device-app/ios-swift/umsDemo/ums/Data/Repository/MessageRepository.swift
import Foundation

import Alamofire

class MessageRepository {
    static let shared = MessageRepository()
    
    func getImage(_ message: DataDef.Message, completion: @escaping (Result<Data, AFError>) -> (Void)) {
        guard let imageId = message.imageId else { return }
        
        let request = PushAppProto.PushAppImgDataReq(
            mAccountId: Config.shared.umsAccountId,
            mImgId: imageId,
            mMsgId: message.id
        )
        
        UmsService.shared.getImageData(request) { [weak self] result in
            guard let self = self else { return }
            
            switch result {
            case let .success(response):
                guard response.mResultCode == UmsResult.RESULT_OK else {
                    NSLog("[\(type(of: self))] [\(response.mResultCode)] \(String(describing: response.mComment))")
                    return
                }
                
                guard let imageData = Data(base64Encoded: response.mImgData) else { return }
                completion(.success(imageData))
            case let .failure(error):
                NSLog("[\(type(of: self))] \(error)")
                
                completion(.failure(error))
                return
            }
        }
    }
    
    func getDeliveryInfo(_ message: DataDef.Message, completion: @escaping (Result<DataDef.Message, AFError>) -> (Void)) {
        let diffDays: Int = Calendar.current.dateComponents([.day], from: message.sentAt, to: Date()).day ?? 0
        
        if diffDays >= 7 {
            completion(.success(message))
            return
        }
        
        let user = UserLocalDataSource.shared.user!
        
        let request = PushAppProto.PushAppMsgInfoReq(
            mAccountId: Config.shared.umsAccountId,
            mPhoneNum: user.phoneNumber,
            mDeviceRegId: user.regId,
            mMsgId: message.id
        )
        
        UmsService.shared.getMessageInfo(request) { [weak self] result in
            guard let self = self else { return }
            
            switch result {
            case let .success(response):
                guard response.mResultCode == UmsResult.RESULT_OK else {
                    NSLog("[\(type(of: self))] [\(response.mResultCode)] \(String(describing: response.mComment))")
                    return
                }
                
                let newMessage = MessageLocalDataSource.shared.updateDeliveryInfo(message.id,
                                                                                  mAlimtalkSendTime: response.mAlimtalkSendTime,
                                                                                  mMunjaSendTime: response.mMunjaSendTime,
                                                                                  mMunjaType: response.mMunjaType)!
                completion(.success(newMessage))
            case let .failure(error):
                NSLog("[\(type(of: self))] \(error)")
                completion(.failure(error))
            }
        }
    }
    
    func markAsRead(_ message: DataDef.Message) {
        sendReadRequest(message)
        MessageLocalDataSource.shared.markAsRead(message.id)
    }
    
    func markAllAsRead() {
        for message in MessageLocalDataSource.shared.messages.value {
            sendReadRequest(message)
        }
        
        MessageLocalDataSource.shared.markAllAsRead()
    }
    
    func sendReadRequest(_ message: DataDef.Message) {
        let user = UserLocalDataSource.shared.user!
        let request = PushAppProto.PushAppMsgReadNoti(mAccountId: Config.shared.umsAccountId,
                                                      mPhoneNum: user.phoneNumber,
                                                      mDeviceRegId: user.regId,
                                                      mMsgId: message.id)
        
        UmsService.shared.setMessageRead(request) { [weak self] result in
            guard let self = self else { return }
            
            if case .failure(let error) = result {
                NSLog("[\(type(of: self))] \(error)")
            }
        }
    }
}

File: device-app/android/Java/umsDemo/app/src/main/java/com/rationalowl/umsdemo/data/repository/MessageRepository.java
package com.rationalowl.umsdemo.data.repository;

import android.util.Base64;
import android.util.Log;

import com.rationalowl.bridgeUms.PushAppProto;
import com.rationalowl.bridgeUms.UmsResult;
import com.rationalowl.umsdemo.api.UmsService;
import com.rationalowl.umsdemo.data.Config;
import com.rationalowl.umsdemo.data.DataCallback;
import com.rationalowl.umsdemo.data.DataDef;
import com.rationalowl.umsdemo.data.datasource.MessageLocalDataSource;
import com.rationalowl.umsdemo.data.datasource.UserLocalDataSource;

import java.io.IOException;
import java.util.Date;
import java.util.concurrent.TimeUnit;

public class MessageRepository {
    private static final String TAG = "MessageRepository";

    private static MessageRepository instance;

    public static synchronized MessageRepository getInstance() {
        if (instance == null) {
            instance = new MessageRepository();
        }

        return instance;
    }

    public void getImage(DataDef.Message message, DataCallback<byte[]> callback) {
        final String imageId = message.getImageId();
        if (imageId == null) return;

        final PushAppProto.PushAppImgDataReq request = new PushAppProto.PushAppImgDataReq();
        request.mAccountId = Config.getInstance().getUmsAccountId();
        request.mImgId = imageId;
        request.mMsgId = message.getId();

        UmsService.getInstance().getImageData(request, new DataCallback<PushAppProto.PushAppImgDataRes>() {
            @Override
            public void onResponse(PushAppProto.PushAppImgDataRes response) {
                final int resultCode = response.mResultCode;

                if (resultCode != UmsResult.RESULT_OK) {
                    Log.e(TAG, "[" + resultCode + "] " + response.mComment);
                    return;
                }

                final byte[] imageData = Base64.decode(response.mImgData, Base64.DEFAULT);
                callback.onResponse(imageData);
            }

            @Override
            public void onFailure(IOException e) {
                callback.onFailure(e);
            }
        });
    }

    public void getDeliveryInfo(DataDef.Message message, DataCallback<DataDef.Message> callback) {
        final long diffInMills = Math.abs(message.getSentAt().getTime() - new Date().getTime());
        final long diffDays = TimeUnit.DAYS.convert(diffInMills, TimeUnit.MILLISECONDS);

        if (diffDays >= 7) {
            callback.onResponse(message);
            return;
        }

        final DataDef.User user = UserLocalDataSource.getInstance().getUser();

        final PushAppProto.PushAppMsgInfoReq request = new PushAppProto.PushAppMsgInfoReq();
        request.mAccountId = Config.getInstance().getUmsAccountId();
        request.mPhoneNum = user.getPhoneNumber();
        request.mDeviceRegId = user.getRegId();
        request.mMsgId = message.getId();

        UmsService.getInstance().getMessageInfo(request, new DataCallback<PushAppProto.PushAppMsgInfoRes>() {
            @Override
            public void onResponse(PushAppProto.PushAppMsgInfoRes response) {
                final int resultCode = response.mResultCode;

                if (resultCode != UmsResult.RESULT_OK) {
                    Log.e(TAG, "[" + resultCode + "] " + response.mComment);
                    return;
                }

                final DataDef.Message newMessage = MessageLocalDataSource.getInstance().updateDeliveryInfo(message.getId(), response.mAlimtalkSendTime, response.mMunjaSendTime, response.mMunjaType);
                callback.onResponse(newMessage);
            }

            @Override
            public void onFailure(IOException e) {
                MessageLocalDataSource.getInstance().save();
                callback.onResponse(message);
            }
        });
    }

    public void setAsRead(DataDef.Message message) {
        sendReadRequest(message);
        MessageLocalDataSource.getInstance().markAsRead(message.getId());
    }

    public void setAllAsRead() {
        for (final DataDef.Message message : MessageLocalDataSource.getInstance().getMessages()) {
            sendReadRequest(message);
        }

        MessageLocalDataSource.getInstance().markAllAsRead();
    }

    private void sendReadRequest(DataDef.Message message) {
        final PushAppProto.PushAppMsgReadNoti request = new PushAppProto.PushAppMsgReadNoti();
        final DataDef.User user = UserLocalDataSource.getInstance().getUser();
        request.mAccountId = Config.getInstance().getUmsAccountId();
        request.mPhoneNum = user.getPhoneNumber();
        request.mDeviceRegId = user.getRegId();
        request.mMsgId = message.getId();

        UmsService.getInstance().setMessageRead(request);
    }
}
